## **Notebook 12 — RAG + Long/Short Term Memory (End to End)**
Integrates the full production RAG pipeline with a persistent memory system.

- Document: data/docs/md_al_amin_resume.pdf
- STM: LangGraph PostgresSaver checkpointer — per-thread conversation history, summarised at 6 messages
- LTM: LangGraph PostgresStore — persistent user profile facts across all threads
- The LLM answers from the resume AND from what it remembers about the user across sessions.

# ***PART 1 — Imports & Model Loading***

## **Cell 1 — install + imports**
- **Why:** All libraries needed for the full pipeline in one place.
- **langgraph-checkpoint-postgres:** provides PostgresSaver (STM) and PostgresStore (LTM). These are official LangGraph persistence backends — they store conversation checkpoints and memory namespaces directly in your existing PostgreSQL database. No separate vector database needed for memory.


In [1]:
# Install if needed
# !pip install langgraph-checkpoint-postgres psycopg[binary] psycopg[pool]
# !pip install FlagEmbedding bm25s docling langchain-openai python-dotenv

import os, json, re, uuid, operator
from pathlib import Path
from typing import TypedDict, Annotated, Optional
from dataclasses import dataclass, field as dc_field
from collections import Counter
from datetime import datetime
from dotenv import load_dotenv

import numpy as np
import psycopg2
from pgvector.psycopg2 import register_vector
import bm25s

from docling.document_converter import DocumentConverter
from FlagEmbedding import BGEM3FlagModel, FlagReranker
from langchain_openai import ChatOpenAI
from langchain_core.messages import (
    SystemMessage, HumanMessage, AIMessage, RemoveMessage
)

# LangGraph core
from langgraph.graph import StateGraph, END
from langgraph.checkpoint.postgres import PostgresSaver
from langgraph.store.postgres import PostgresStore

load_dotenv()

DATABASE_SYNC = os.getenv("DATABASE_URL_SYNC")
OPENAI_KEY    = os.getenv("OPENAI_API_KEY")
INDEX_DIR     = Path("../data/bm25_index_resume")
INDEX_DIR.mkdir(parents=True, exist_ok=True)

print("Imports OK")
print(f"DB  : {DATABASE_SYNC[:10]}...")
print(f"Key : {OPENAI_KEY[:15]}...")

Imports OK
DB  : postgresql...
Key : sk-proj-Pm8wgjk...


## **Cell 2 — load AI models**
- **Why load all models here:** BGE-M3 (~2.3GB) and the reranker (~1.1GB) take 20-30s to load. Loading once at the top means every subsequent cell is instant.
- **GPT-4o-mini for everything:** We use gpt-4o-mini for HyDE generation, LTM fact extraction, STM summarisation, and final answer generation. It is fast, cheap, and accurate enough for all four tasks.

In [2]:
MODEL_DIR = Path("../models")

print("Loading BGE-M3 embedder...")
embed_model = BGEM3FlagModel(
    "BAAI/bge-m3", use_fp16=False,
    cache_dir=str(MODEL_DIR / "bge-m3")
)
print("  BGE-M3 ready")

print("Loading BGE reranker...")
rerank_model = FlagReranker(
    "BAAI/bge-reranker-v2-m3", use_fp16=False,
    cache_dir=str(MODEL_DIR / "bge-reranker")
)
print("  Reranker ready")

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0.1)
print("  LLM ready (gpt-4o-mini)")

print("\nAll models loaded OK")

Loading BGE-M3 embedder...


Fetching 30 files:   0%|          | 0/30 [00:00<?, ?it/s]

  BGE-M3 ready
Loading BGE reranker...
  Reranker ready
  LLM ready (gpt-4o-mini)

All models loaded OK


## ***Cell 3 — fresh DB tables for this notebook**
- **Why fresh tables:** Isolated from previous notebooks so you can re-run from scratch without corrupting earlier experiments.
chunks_resume + documents_resume: Same schema as before but with suffix to avoid collisions.
- **pgvector HNSW index:** Created immediately after table creation since we have few chunks (resume is small). In production with large corpora, create the HNSW index AFTER bulk loading all data for performance.

In [3]:
conn_str = DATABASE_SYNC.replace("postgresql+psycopg2://", "postgresql://")
conn     = psycopg2.connect(conn_str)
register_vector(conn)
cur      = conn.cursor()

# Drop and recreate for clean start
cur.execute("DROP TABLE IF EXISTS chunks_resume   CASCADE;")
cur.execute("DROP TABLE IF EXISTS documents_resume CASCADE;")

cur.execute("""
    CREATE TABLE documents_resume (
        id         UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        filename   TEXT NOT NULL,
        status     TEXT DEFAULT 'ready',
        created_at TIMESTAMPTZ DEFAULT now()
    );
""")

cur.execute("""
    CREATE TABLE chunks_resume (
        id             UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        document_id    UUID REFERENCES documents_resume(id) ON DELETE CASCADE,
        content        TEXT NOT NULL,
        parent_content TEXT,
        embedding      vector(1024),
        metadata       JSONB NOT NULL DEFAULT '{}',
        created_at     TIMESTAMPTZ DEFAULT now()
    );
""")

cur.execute("""
    CREATE INDEX idx_resume_embedding
    ON chunks_resume
    USING hnsw (embedding vector_cosine_ops)
    WITH (m=16, ef_construction=64);
""")

cur.execute("""
    CREATE INDEX idx_resume_metadata
    ON chunks_resume USING gin(metadata);
""")

conn.commit()
print("Tables created: documents_resume, chunks_resume")
print("Indexes created: HNSW (embedding), GIN (metadata)")

Tables created: documents_resume, chunks_resume
Indexes created: HNSW (embedding), GIN (metadata)


## **Cell 4 — parse resume PDF with Docling**
- **Why Docling over PyPDF2:** Your resume has structured sections — Education, Experience, Projects, Skills. Docling preserves these as markdown headers which become section metadata on every chunk. PyPDF2 would flatten them into one block of text.
- **Path:**  data/docs/md_al_amin_resume.pdf — the file you uploaded.

In [4]:
PDF_PATH = Path("../data/docs/md_al_amin_resume.pdf")

if not PDF_PATH.exists():
    print(f"PDF not found at {PDF_PATH}")
    print("Make sure md_al_amin_resume.pdf is in data/docs/")
else:
    print(f"Found: {PDF_PATH}")
    print(f"Size : {PDF_PATH.stat().st_size / 1024:.1f} KB")

converter = DocumentConverter()
print("\nParsing with Docling...")
result   = converter.convert(str(PDF_PATH))
markdown = result.document.export_to_markdown()

print(f"Parsed successfully")
print(f"Characters: {len(markdown)}")
print(f"\nMarkdown preview:")
print(markdown[:600])

Found: ../data/docs/md_al_amin_resume.pdf
Size : 117.2 KB

Parsing with Docling...


[INFO] 2026-05-02 12:16:24,064 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-02 12:16:24,073 [RapidOCR] download_file.py:60: File exists and is valid: /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-05-02 12:16:24,074 [RapidOCR] main.py:53: Using /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-05-02 12:16:24,138 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-02 12:16:24,141 [RapidOCR] download_file.py:60: File exists and is valid: /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-05-02 12:16:24,141 [RapidOCR] main.py:53: Using /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site

Parsed successfully
Characters: 3766

Markdown preview:
## Education

## North South University

Bachelor of Science in Computer Science and Engineering

Jan 18 - Apr 25

CGPA: 3.17 (86%)

|

Coursework: Data Structures, Algorithms, DBMS, ML, DL, NLP, Operating Systems

## Technical Skills

Languages : Python | C/C++ | SQL | JavaScript

GenAI &amp; ML : PyTorch | TensorFlow | LLM | LangChain | LangGraph | vLLM | MCP | Prompt and Context Engineering | Embedding | RAG | Multi-Model RAG | PEFT(LoRA, QLoRA) | VectorDB | OpenAI API | Whisper | STT Web &amp; DevOps : FastAPI, Docker, GitLab CI/CD, Redis, NGiNX, PostgreSQL, RunPod, Playwright, Microservic


## **Cell 5 — chunking functions**
- **Why smaller chunks for a resume:** A resume has dense, structured content — each bullet point is a complete fact. We use child_size=2 sentences and parent_size=4 sentences. Larger windows would blend Skills with Experience which hurts retrieval precision.
- **Section detection:** The detect_section function walks backwards through the markdown lines to find the nearest ## header. Every chunk gets tagged with its section — "Experience", "Projects", "Skills" etc. This enables filtered queries like "only search within Skills section".

In [5]:
def detect_section(lines: list[str], up_to: int) -> str:
    """Walk backwards from current position to find nearest ## header."""
    for line in reversed(lines[:up_to]):
        line = line.strip()
        if line.startswith("## "):
            return line.lstrip("#").strip()
        if line.startswith("# "):
            return line.lstrip("#").strip()
    return "General"

def split_sentences(text: str) -> list[str]:
    """Split into sentences, skip headers and short fragments."""
    sentences = []
    for line in text.split("\n"):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        parts = re.split(r'(?<=[.!?])\s+', line)
        for part in parts:
            if len(part.strip()) > 15:
                sentences.append(part.strip())
    return sentences

def build_chunks(markdown: str, filename: str,
                 child_size: int = 2,
                 parent_size: int = 4,
                 overlap: int = 1) -> list[dict]:
    """
    Build parent-child chunks from markdown text.

    child_size=2  — 2 sentences per child (precise retrieval embedding)
    parent_size=4 — 4 sentences per parent (full context for LLM)
    overlap=1     — 1 sentence shared between adjacent chunks
                    prevents answers that span chunk boundaries from being missed

    Returns list of dicts with content, parent_content, section, metadata.
    """
    sentences = split_sentences(markdown)
    lines     = markdown.split("\n")
    chunks    = []
    idx       = 0
    i         = 0

    while i < len(sentences):
        parent_sents   = sentences[i : i + parent_size]
        parent_content = " ".join(parent_sents)
        section        = detect_section(lines, min(i * 2, len(lines) - 1))

        j = 0
        while j < len(parent_sents):
            child_content = " ".join(parent_sents[j : j + child_size])
            if len(child_content.strip()) > 15:
                chunks.append({
                    "content"       : child_content,
                    "parent_content": parent_content,
                    "section"       : section,
                    "metadata"      : {
                        "filename"   : filename,
                        "section"    : section,
                        "chunk_index": idx,
                    }
                })
                idx += 1
            j += child_size - overlap
        i += parent_size - overlap

    return chunks

chunks = build_chunks(markdown, "md_al_amin_resume.pdf")
print(f"Total chunks: {len(chunks)}")
print(f"\nSection distribution:")
for section, count in Counter(c["section"] for c in chunks).most_common():
    print(f"  {section:<35} {count:>3} chunks")
print(f"\nSample chunk:")
print(f"  Child  : {chunks[0]['content'][:100]}")
print(f"  Parent : {chunks[0]['parent_content'][:100]}")

Total chunks: 41

Section distribution:
  North South University                8 chunks
  Genuine Technology &amp; Research Ltd   8 chunks
  Jr. Generative AI Engineer            8 chunks
  1. SynapseAI High-Performance Memory Agent | FastAPI, LangGraph, Redis, PostgreSQL   8 chunks
  General                               4 chunks
  Technical Skills                      4 chunks
  3. Agentic AI Math Tutor with Adaptive Prompting | LangGraph, ChromaDB, Llama-4 [Capstone Proj-Apr 2025]   1 chunks

Sample chunk:
  Child  : Bachelor of Science in Computer Science and Engineering CGPA: 3.17 (86%)
  Parent : Bachelor of Science in Computer Science and Engineering CGPA: 3.17 (86%) Coursework: Data Structures


## **Cell 6 — embed chunks + store in pgvector**
- **Why batch_size=16:** Resume has ~38 chunks. A single batch is fine. For larger documents use batch_size=32 and loop.
- **What gets stored:** The child chunk content, parent chunk content, the 1024-dim BGE-M3 dense embedding, and JSONB metadata. The embedding is what pgvector searches. The parent content is what the LLM receives.

In [6]:
texts    = [c["content"] for c in chunks]
print(f"Embedding {len(texts)} chunks with BGE-M3...")

output   = embed_model.encode(
    texts,
    batch_size=16,
    return_dense=True,
    return_sparse=False,
)
vecs = output["dense_vecs"]
print(f"Embeddings shape: {vecs.shape}")

# Insert document record
cur.execute(
    "INSERT INTO documents_resume (filename) VALUES (%s) RETURNING id;",
    ("md_al_amin_resume.pdf",)
)
doc_id = cur.fetchone()[0]

# Insert all chunks
for chunk, vec in zip(chunks, vecs):
    cur.execute("""
        INSERT INTO chunks_resume
            (document_id, content, parent_content, embedding, metadata)
        VALUES (%s, %s, %s, %s, %s::jsonb)
    """, (
        doc_id,
        chunk["content"],
        chunk["parent_content"],
        vec.tolist(),
        json.dumps(chunk["metadata"]),
    ))

conn.commit()

cur.execute("SELECT COUNT(*) FROM chunks_resume;")
print(f"Stored {cur.fetchone()[0]} chunks in pgvector")

Embedding 41 chunks with BGE-M3...


pre tokenize: 100%|██████████| 3/3 [00:00<00:00, 361.81it/s]
You're using a XLMRobertaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
Inference Embeddings: 100%|██████████| 3/3 [00:00<00:00,  6.31it/s]


Embeddings shape: (41, 1024)
Stored 41 chunks in pgvector


## **Cell 7 — build BM25s index**

- **Why BM25s on a resume:** Resumes are full of exact terms — "LangGraph", "FastAPI", "CGPA 3.17", "North South University", "July 2025". Dense embeddings may miss these exact strings because they represent rare proper nouns. BM25s catches them reliably.
- **Load chunk_ids from DB:** We fetch IDs in insertion order so the BM25s index array position maps exactly to the correct chunk UUID in pgvector.

In [7]:
chunk_texts = [c["content"] for c in chunks]

cur.execute("SELECT id::text FROM chunks_resume ORDER BY created_at;")
chunk_ids = [r[0] for r in cur.fetchall()]

print(f"Building BM25s index over {len(chunk_texts)} chunks...")
corpus_tokens  = bm25s.tokenize(chunk_texts, stopwords="en")
bm25_retriever = bm25s.BM25()
bm25_retriever.index(corpus_tokens)

bm25_retriever.save(str(INDEX_DIR / "bm25_index"))
with open(INDEX_DIR / "chunk_ids.json",   "w") as f: json.dump(chunk_ids,   f)
with open(INDEX_DIR / "chunk_texts.json", "w") as f: json.dump(chunk_texts, f)

print(f"BM25s index saved — {len(chunk_ids)} entries")
print("Ingestion complete")

Building BM25s index over 41 chunks...


Split strings:   0%|          | 0/41 [00:00<?, ?it/s]

BM25S Count Tokens:   0%|          | 0/41 [00:00<?, ?it/s]

BM25S Compute Scores:   0%|          | 0/41 [00:00<?, ?it/s]

BM25s index saved — 41 entries
Ingestion complete


A# ***RT 3 — Memory System Setup***


## **Cell 8 — setup PostgresSaver (STM) and PostgresStore (LTM)**

- **PostgresSaver (STM):** Stores the conversation message history for each thread. When you pass a thread_id config to the graph, LangGraph automatically saves and restores the message list from this checkpointer. Each thread has its own independent history.
- **PostgresStore (LTM):** Stores arbitrary key-value facts in namespaces. We use namespace ("user", user_id, "profile") for user profile facts. These persist across ALL threads for that user — unlike the checkpointer which is thread-scoped.
- **Connection string format:** PostgresSaver and PostgresStore require the psycopg3 connection string format — postgresql://user:pass@host:port/db.

In [8]:
from psycopg_pool import ConnectionPool
from langgraph.checkpoint.postgres import PostgresSaver
from langgraph.store.postgres import PostgresStore
from functools import partial

# Use the same DB URL pattern from your old notebook
# Your .env has DATABASE_URL_SYNC — convert to psycopg3 format
PG_CONN_STR = DATABASE_SYNC.replace("postgresql+psycopg2://", "postgresql://")

# ── Connection pool — shared by both STM and LTM ─────────────
# max_size=20 — max concurrent connections in the pool
# autocommit=True — required by both PostgresSaver and PostgresStore
pool = ConnectionPool(
    conninfo=PG_CONN_STR,
    max_size=20,
    kwargs={"autocommit": True},
)

# ── STM: PostgresSaver ────────────────────────────────────────
# Pass the pool directly — not from_conn_string
checkpointer = PostgresSaver(pool)
checkpointer.setup()   # creates langgraph_checkpoints table
print("STM checkpointer (PostgresSaver) ready")

# ── LTM: PostgresStore ────────────────────────────────────────
# Same pool — no second connection needed
ltm_store = PostgresStore(pool)
ltm_store.setup()      # creates langgraph_store table
print("LTM store (PostgresStore) ready")

print()
print("Memory tables in PostgreSQL:")
print("  langgraph_checkpoints — STM (conversation history per thread)")
print("  langgraph_store       — LTM (user profile facts across threads)")

STM checkpointer (PostgresSaver) ready
LTM store (PostgresStore) ready

Memory tables in PostgreSQL:
  langgraph_checkpoints — STM (conversation history per thread)
  langgraph_store       — LTM (user profile facts across threads)


## **Cell 9 — MemoryOp schema + LTM manager prompt**

- **Why MemoryOp schema:** Instead of asking the LLM to write free-form memory updates, we define a strict JSON schema with three operations: create, update, delete. The LLM must respond with a list of these operations. This makes memory management deterministic and auditable — you can see exactly what facts were added, changed, or removed after every message.
- **The LTM manager prompt:** After every user message + assistant response pair, we send both to the LLM with this prompt. It decides which facts (if any) to create, update, or delete in the user's profile. If nothing important was said, it returns an empty list.

In [9]:
from pydantic import BaseModel

class MemoryOp(BaseModel):
    """
    One memory operation — create, update, or delete a fact about the user.

    action : "create"  — store a new fact that was not known before
             "update"  — correct or extend an existing fact
             "delete"  — remove a fact that is no longer true

    key    : short identifier for the fact, e.g. "name", "role", "university"
    value  : the fact content (empty string for delete operations)
    reason : brief explanation of why this operation is needed
    """
    action : str    # "create" | "update" | "delete"
    key    : str    # e.g. "name", "role", "location", "goal"
    value  : str    # the fact value
    reason : str    # why this operation

LTM_MANAGER_PROMPT = """You are a memory manager for an AI assistant.

Your job: after reading the conversation exchange below, decide what facts
about the USER should be stored, updated, or deleted in their long-term profile.

Rules:
- Only store facts the user explicitly stated about themselves
- Do NOT store opinions, questions, or temporary states
- Do NOT store facts from the document (resume) — only facts the user told you
- If the user corrects a previous statement, update the old fact
- If nothing important was said, return an empty list
- Keep fact values short and factual (under 30 words)

User message : {user_message}
Assistant response: {assistant_response}

Return a JSON array of memory operations. Each must have:
  action : "create" | "update" | "delete"
  key    : short snake_case identifier
  value  : the fact (empty for delete)
  reason : why this operation

Return [] if nothing should be stored. Return only valid JSON, no other text."""

STM_SUMMARY_PROMPT = """Summarise this conversation in 3-5 sentences.
Focus on: what the user asked about, what topics were covered, any important
context that would help answer future questions in this conversation.
Be concise. Do not include greetings or filler.

Conversation:
{messages}

Summary:"""

print("MemoryOp schema defined")
print("LTM_MANAGER_PROMPT defined")
print("STM_SUMMARY_PROMPT defined")

MemoryOp schema defined
LTM_MANAGER_PROMPT defined
STM_SUMMARY_PROMPT defined


## **Cell 10 — LTM helper functions**

- **get_ltm_profile:** Loads all stored facts for a user from PostgresStore. Returns a formatted string like "name: Al Amin | role: AI Engineer | goal: PhD scholarship". This string is injected into the generation prompt so the LLM knows who it is talking to.
- **apply_memory_ops:** Executes a list of MemoryOp objects against PostgresStore. Create and Update both call store.put() — pgstore is upsert-by-default. Delete calls store.delete(). This function is called by the update_memory node after every response.


In [10]:
def get_ltm_profile(store, user_id: str) -> str:
    """
    Load all LTM facts for a user and return as a formatted string.

    Queries PostgresStore namespace ("user", user_id, "profile").
    Returns formatted string for injection into generation prompt.
    Returns empty string if no facts exist yet (new user).

    Example output:
        "name: Al Amin | role: Jr. AI Engineer | goal: PhD scholarship"
    """
    namespace = ("user", user_id, "profile")
    try:
        items = store.search(namespace)
        if not items:
            return ""
        facts = [f"{item.key}: {item.value}" for item in items
                 if isinstance(item.value, str)]
        return " | ".join(facts)
    except Exception:
        return ""

def apply_memory_ops(store, user_id: str, ops: list[MemoryOp]) -> int:
    """
    Execute a list of MemoryOp objects against PostgresStore.

    create / update → store.put(namespace, key, value)
    delete          → store.delete(namespace, key)

    Returns count of operations applied.
    """
    namespace = ("user", user_id, "profile")
    applied   = 0

    for op in ops:
        try:
            if op.action in ("create", "update"):
                store.put(namespace, op.key, op.value)
                applied += 1
            elif op.action == "delete":
                store.delete(namespace, op.key)
                applied += 1
        except Exception as e:
            print(f"  [memory] op failed: {op.action} {op.key} — {e}")

    return applied

def extract_memory_ops(user_message: str, assistant_response: str) -> list[MemoryOp]:
    """
    Ask gpt-4o-mini to extract memory operations from a conversation turn.
    Returns list of MemoryOp objects (may be empty if nothing to store).
    """
    prompt   = LTM_MANAGER_PROMPT.format(
        user_message       = user_message,
        assistant_response = assistant_response,
    )
    response = llm.invoke([HumanMessage(content=prompt)])
    raw      = response.content.strip()

    # Strip markdown code fences if present
    if raw.startswith("```"):
        raw = re.sub(r"^```[a-z]*\n?", "", raw)
        raw = re.sub(r"\n?```$",        "", raw)

    try:
        ops_data = json.loads(raw)
        return [MemoryOp(**op) for op in ops_data if isinstance(op, dict)]
    except Exception:
        return []

print("LTM helper functions defined:")
print("  get_ltm_profile()      — loads user profile from PostgresStore")
print("  apply_memory_ops()     — executes create/update/delete ops")
print("  extract_memory_ops()   — LLM extracts ops from conversation turn")

LTM helper functions defined:
  get_ltm_profile()      — loads user profile from PostgresStore
  apply_memory_ops()     — executes create/update/delete ops
  extract_memory_ops()   — LLM extracts ops from conversation turn


## **Cell 11 — STM helper functions**

- **get_stm_summary:** Reads the current conversation summary from the LangGraph checkpointer state. LangGraph stores messages as a list in the checkpoint. We look for a special summary message we injected previously.

- **summarise_and_trim:** When message count exceeds 6, we: (1) ask gpt-4o-mini to summarise the full conversation, (2) generate RemoveMessage objects for all old messages, (3) return the summary text + remove operations. The LangGraph state graph applies the removes automatically via the operator.add annotation on the messages field, which handles RemoveMessage correctly.

In [ ]:
STM_WINDOW  = 6    # trigger summarisation when message count exceeds this
STM_KEEP    = 2    # keep this many recent messages after summarisation

def get_messages_from_state(state: dict) -> list:
    """Extract the messages list from the LangGraph state."""
    return state.get("messages", [])

def should_summarise(messages: list) -> bool:
    """Return True if message count exceeds STM_WINDOW."""
    return len(messages) > STM_WINDOW

def summarise_messages(messages: list) -> str:
    """
    Ask gpt-4o-mini to summarise the conversation history.
    Called when message count exceeds STM_WINDOW.

    Returns a plain text summary string.
    """
    # Format messages for the prompt
    formatted = []
    for msg in messages:
        if isinstance(msg, HumanMessage):
            formatted.append(f"User: {msg.content}")
        elif isinstance(msg, AIMessage):
            formatted.append(f"Assistant: {msg.content}")
    conversation_text = "\n".join(formatted)

    prompt   = STM_SUMMARY_PROMPT.format(messages=conversation_text)
    response = llm.invoke([HumanMessage(content=prompt)])
    return response.content.strip()

def build_trim_ops(messages: list, keep_last: int = STM_KEEP) -> list:
    """
    Build a list of RemoveMessage operations to trim old messages.
    Keeps the last `keep_last` messages.
    Returns list of RemoveMessage objects.

    LangGraph handles RemoveMessage via the operator.add annotation:
    when the state update contains RemoveMessage, LangGraph removes
    that message from the list instead of appending.
    """
    if len(messages) <= keep_last:
        return []
    messages_to_remove = messages[:-keep_last]
    return [RemoveMessage(id=m.id) for m in messages_to_remove
            if hasattr(m, "id") and m.id]

print("STM helper functions defined:")
print("  should_summarise()   — checks if message count > STM_WINDOW")
print("  summarise_messages() — LLM summarises conversation history")
print("  build_trim_ops()     — builds RemoveMessage list for trimming")